### PR TITLE
tensorflow-lite: v2 migration + unvendor more deps

### PR DESCRIPTION
**Pull request by prince-chrismc**
Specify library name and version:  **tensorflow-lite/2.9.1**

This is also a good place to share with all of us **why you are submitting this PR** (specially if it is a new addition to ConanCenter): is it a dependency of other libraries you want to package? Are you the author of the library? Thanks!

I was plaything with this after review #11477 and notice a few issue in which deps where being handled. This approach allowed me to find more issue but it's still problematic when adding new versions.

I need to retest windows since i've rebased

---

- [x] I've read the [guidelines](https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md) for contributing.
- [x] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [x] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [x] I've tried at least one configuration locally with the [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

**Comment by prince-chrismc** [comment]
Self reminder to run the v1 test package locally you can use

```
conan test recipes/tensorflow-lite/all/test_v1_package/conanfile.py tensorflow-lite/2.9.1@
```

I need to write the docs for this

File: recipes/tensorflow-lite/all/patches/dependencies_2_9.patch [hunk 1]
```diff
@@ -1,81 +0,0 @@
...
- find_package(cpuinfo REQUIRED)  #CPUINFO is used by XNNPACK and RUY library
- find_package(ruy REQUIRED)
-@@ -162,6 +155,10 @@ set(TFLITE_TARGET_PUBLIC_OPTIONS "")
- set(TFLITE_TARGET_PRIVATE_OPTIONS "")
- # Additional library dependencies based upon enabled features.
- set(TFLITE_TARGET_DEPENDENCIES "")
-+if (NOT CMAKE_SYSTEM_PROCESSOR OR CMAKE_SYSTEM_PROCESSOR MATCHES "x86")
-+        find_package(NEON_2_SSE REQUIRED)
-+        list(APPEND TFLITE_TARGET_DEPENDENCIES NEON_2_SSE::NEON_2_SSE)
-+endif()
```
Comment (prince-chrismc) [inline recipes/tensorflow-lite/all/patches/dependencies_2_9.patch:45]: Forgot about M1 ... I'll open a PR for this as well

Comment (prince-chrismc) [reply recipes/tensorflow-lite/all/patches/dependencies_2_9.patch:45]: https://github.com/tensorflow/tensorflow/pull/57548

**Comment by prince-chrismc** [comment]
🤞 this should finally pass!

File: recipes/tensorflow-lite/all/test_package/conanfile.py [hunk 2]
```diff
@@ -1,18 +1,26 @@
...
+        cmake_layout(self)
 
     def build(self):
         cmake = CMake(self)
         cmake.configure()
         cmake.build()
 
     def test(self):
-        if not tools.cross_building(self):
+        if not build.cross_building(self):
```
Comment (uilianries) [inline recipes/tensorflow-lite/all/test_package/conanfile.py:23]:
```suggestion
        if build.can_run(self):
```
Prefer can_run. It incorporates `cross_building`, but it's managed by conan conf too. 
https://docs.conan.io/en/latest/reference/conanfile/tools/build.html#conan-tools-build-can-run

**Review by uilianries — COMMENTED** [review]
Nice job, only a little improvement.

**Comment by prince-chrismc** [comment]
Thanks 😊

**Review by uilianries — APPROVED** [review]
LGTM

**Comment by prince-chrismc** [comment]
@SpaceIm @toge @ericLemanissier 👋 looking for a review please!

File: recipes/tensorflow-lite/all/test_v1_package/conanfile.py [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+    generators = "cmake", "cmake_find_package_multi"
+
+    def build(self):
+        cmake = CMake(self)
+        cmake.configure()
+        cmake.build()
+
+    def test(self):
+        if not tools.cross_building(self):
+            model_path = os.path.join(self.source_folder, "..", "test_package", "model.tflite")
```
Comment (toge) [inline recipes/tensorflow-lite/all/test_v1_package/conanfile.py:16]:
```suggestion
            model_path = os.path.join(self.source_folder, os.pardir, "test_package", "model.tflite")
```

Please use `os.pardir`.

File: recipes/tensorflow-lite/all/conanfile.py [hunk 4]
```diff
@@ -82,54 +70,83 @@ def requirements(self):
         self.requires("fft/cci.20061228")
         self.requires("flatbuffers/2.0.5")
         self.requires("gemmlowp/cci.20210928")
+        self.requires("opencl-headers/2022.05.18")
+        self.requires("vulkan-headers/1.3.221")
         if self.settings.arch in ("x86", "x86_64"):
             self.requires("intel-neon2sse/cci.20210225")
-        self.requires("ruy/cci.20220628")
+        if self.options.with_ruy or True: # Upstream always calls find_package
```
Comment (jgsogo) [inline recipes/tensorflow-lite/all/conanfile.py:77]:
```suggestion
        if self.options.with_ruy or True: # FIXME: Upstream always calls find_package
```

Maybe someone provides a patch for the build-system

**Review by talyz — COMMENTED** [review]
This seems to raise the complexity of the recipe significantly, especially when it comes to the dependency patch. It also removes the `2.6.x` series, which is probably fine, but it should at least be mentioned.

I have a much smaller changeset which essentially does the same thing in #12390. It's currently missing some of the fixes and updates made here, but it currently looks to be more maintainable and keeps the `2.6.x` series for now.

It doesn't matter too much which PR we go with for me - both need some work before merging IMO:
- this PR should be simplified
  - the patches can be almost entirely reverted
  - the `_POPULATED` variables can be removed (except for `clog`)
  - `OverridableFetchContent` and `FetchContent` should be disabled
- my PR should include: 
  - test updates
  - GCC rev update
  - misc conan v2 fixes

**Comment by prince-chrismc** [comment]
I agree it's not perfect.

> This seems to raise the complexity of the recipe significantly

Yes, but that's partly the fault of the upstream, I want to ensure the deps come from Conan and this is the mechanism that's provide. Personally I think it's best to follow upstream instead of writing our own build scripts.

I would argue the old patches were far too aggressive - especially since they did not correctly fix all the deps.
If you try to unvedor the 3 I'm adding from Conan + the two that are still missing. Its way more error prone with how the build scripts are designed.

You are absolutely right this pays the price in the recipe. Instead of the patches.

> the _POPULATED variables can be removed

Thats the only way to ensure the deps come from Conan 🤔 
Right now the recipe is building a bunch of deps and causing conflicts. 

> drops 2.6

I specifically designed the patches to work with 2.10 thats coming shortly (which 2.6 was too different to include) 

And they will always be available in ConanCenter just no longer updated.

If you look at main https://github.com/tensorflow/tensorflow/blob/2b17986c1ac5b7d6a5eef37fb617d15fe13c0611/tensorflow/lite/CMakeLists.txt#L141

You will see they updated the dependencies to use the correct official names

This will remove all the patches from my testing + my two PRs that are open 🤞 they will get included

**Comment by talyz** [comment]
> Personally I think it's best to follow upstream instead of writing our own build scripts.

Yeah, I'm talking about the current state of the recipe, where we only do minimal patching and don't write our own build script.

> I would argue the old patches were far too aggressive - especially since they did not correctly fix all the deps. If you try to unvedor the 3 I'm adding from Conan + the two that are still missing. Its way more error prone with how the build scripts are designed.

Aggressive how? They do far less than the patches here. Look at my PR and compare. I also introduced a mechanism so that we can be sure no additional dependencies are fetched by the build script.
 
> Thats the only way to ensure the deps come from Conan thinking Right now the recipe is building a bunch of deps and causing conflicts.

See my PR for a better solution.

> I specifically designed the patches to work with 2.10 thats coming shortly (which 2.6 was too different to include)
> 
> And they will always be available in ConanCenter just no longer updated.

Fair enough, I just thought maybe the description should mention it. 2.10 is out now, btw :)

**Comment by prince-chrismc** [comment]
@talyz Thank you for challenging my approach.

I cleaned my system and redid my build and discovered I had https://github.com/tensorflow/tensorflow/blob/d8ce9f9c301d021a69953134185ab728c1c248d3/tensorflow/lite/CMakeLists.txt#L228 `TFLITE_ENABLE_GPU=True` which obvious changed thing since it pull in a bunch of stuff

If I set that off, then your approach is definitely more reasonable.

**Comment by prince-chrismc** [comment]
> Fair enough, I just thought maybe the description should mention it. 2.10 is out now, btw :)

🚀 I just noticed there's also 2.9.2 lol we will be busy!

**Comment by prince-chrismc** [comment]
The other PR went in since it was better

**Pull request closed without merging**